cloud: consolidate ReadFile APIs [draft WIP]

RaduBerinde opened the pull request:
This change consolidates the `ReadFile` and `ReadFileAt` APIs in `cloud.ExternalStorage`. We use a `ReadOptions` struct to optionally specify the offset or that we don't care about the size. This will allow us to add more options without large code changes.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103845)
<!-- Reviewable:end -->

RaduBerinde committed:
cloud: consolidate ReadFile APIs

This change consolidates the `ReadFile` and `ReadFileAt` APIs in
`cloud.ExternalStorage`. We use a `ReadOptions` struct to optionally
specify the offset or that we don't care about the size. This will
allow us to add more options without large code changes.

Epic: none
Release note: None

RaduBerinde committed:
cloud: add read LimitHint, implement for S3 and GCS

Epic: none
Release note: None

RaduBerinde committed:
storage: set LimitHint for pebble object reads

Epic: none
Release note: None
